File: features/auth/fastify/files/src/modules/auth/mfa.ts
import { randomInt } from 'node:crypto';
import { authenticator } from 'otplib';
import { encryptString, decryptString } from '../../lib/crypto.js';
import { hashPassword, verifyPassword } from './password.js';

const ISSUER = process.env.MFA_ISSUER ?? 'projx';
const RECOVERY_CODE_COUNT = 10;
const RECOVERY_CODE_ALPHABET = 'ABCDEFGHJKLMNPQRSTUVWXYZ23456789';
const TOTP_WINDOW = 3;

authenticator.options = { window: TOTP_WINDOW };

export function generateSecret(): string {
  return authenticator.generateSecret(20);
}

export function buildOtpauthUrl(email: string, secret: string): string {
  return authenticator.keyuri(email, ISSUER, secret);
}

export function verifyTotp(code: string, secret: string): boolean {
  return authenticator.verify({ token: code.trim(), secret });
}

export function generateRecoveryCodes(count = RECOVERY_CODE_COUNT): string[] {
  const codes: string[] = [];
  for (let i = 0; i < count; i += 1) {
    codes.push(`${pickChars(4)}-${pickChars(4)}`);
  }
  return codes;
}

function pickChars(length: number): string {
  let out = '';
  for (let i = 0; i < length; i += 1) {
    const idx = randomInt(RECOVERY_CODE_ALPHABET.length);
    out += RECOVERY_CODE_ALPHABET[idx];
  }
  return out;
}

function denormalize(code: string): string {
  const stripped = code.trim().toUpperCase().replace(/\s+/g, '').replace(/-/g, '');
  return `${stripped.slice(0, 4)}-${stripped.slice(4)}`;
}

export async function hashRecoveryCodes(codes: string[]): Promise<string[]> {
  return Promise.all(codes.map((code) => hashPassword(denormalize(code))));
}

export async function matchRecoveryCode(input: string, hashes: string[]): Promise<number> {
  const normalized = denormalize(input);
  for (let i = 0; i < hashes.length; i += 1) {
    if (await verifyPassword(normalized, hashes[i])) return i;
  }
  return -1;
}

export function encryptRecoveryCodes(hashes: string[]): string {
  return encryptString(JSON.stringify(hashes));
}

export function decryptRecoveryCodes(enc: string | null | undefined): string[] {
  if (!enc) return [];
  try {
    const parsed = JSON.parse(decryptString(enc)) as unknown;
    return Array.isArray(parsed) ? (parsed as string[]) : [];
  } catch {
    return [];
  }
}

export function encryptSecret(secret: string): string {
  return encryptString(secret);
}

export function decryptSecret(enc: string): string {
  return decryptString(enc);
}

export function isMfaLocked(lockedUntil: Date | null | undefined): boolean {
  if (!lockedUntil) return false;
  return lockedUntil.getTime() > Date.now();
}

export const MFA_MAX_ATTEMPTS = 5;
export const MFA_LOCKOUT_MS = 15 * 60 * 1000;

File: features/auth/fastapi/files/src/auth/_mfa.py
import os
import secrets
from datetime import UTC, datetime

import pyotp

from src.utils import decrypt_config, encrypt_config

from ._password import hash_password, verify_password

_ISSUER = os.getenv("MFA_ISSUER", "projx")
_RECOVERY_CODE_COUNT = 10
_RECOVERY_CODE_ALPHABET = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"
_TOTP_WINDOW = 3

MFA_MAX_ATTEMPTS = 5
MFA_LOCKOUT_MINUTES = 15


def generate_secret() -> str:
    return pyotp.random_base32(length=32)


def build_otpauth_url(email: str, secret: str) -> str:
    return pyotp.TOTP(secret).provisioning_uri(name=email, issuer_name=_ISSUER)


def verify_totp(code: str, secret: str) -> bool:
    cleaned = code.strip()
    if not cleaned:
        return False
    return pyotp.TOTP(secret).verify(cleaned, valid_window=_TOTP_WINDOW)


def _pick_chars(length: int) -> str:
    return "".join(secrets.choice(_RECOVERY_CODE_ALPHABET) for _ in range(length))


def generate_recovery_codes(count: int = _RECOVERY_CODE_COUNT) -> list[str]:
    return [f"{_pick_chars(4)}-{_pick_chars(4)}" for _ in range(count)]


def _denormalize(code: str) -> str:
    stripped = code.strip().upper().replace(" ", "").replace("-", "")
    return f"{stripped[:4]}-{stripped[4:]}"


def hash_recovery_codes(codes: list[str]) -> list[str]:
    return [hash_password(_denormalize(code)) for code in codes]


def match_recovery_code(input_code: str, hashes: list[str]) -> int:
    normalized = _denormalize(input_code)
    for i, h in enumerate(hashes):
        if verify_password(normalized, h):
            return i
    return -1


def encrypt_recovery_codes(hashes: list[str]) -> str:
    return encrypt_config({"hashes": hashes})


def decrypt_recovery_codes(enc: str | None) -> list[str]:
    if not enc:
        return []
    try:
        data = decrypt_config(enc)
    except Exception:
        return []
    raw = data.get("hashes")
    if not isinstance(raw, list):
        return []
    return [str(h) for h in raw]


def encrypt_secret(secret: str) -> str:
    return encrypt_config({"secret": secret})


def decrypt_secret(enc: str) -> str:
    data = decrypt_config(enc)
    secret = data.get("secret")
    if not isinstance(secret, str):
        raise ValueError("MFA secret payload is malformed")
    return secret


def is_mfa_locked(locked_until: datetime | None) -> bool:
    if locked_until is None:
        return False
    dt = locked_until
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=UTC)
    return dt > datetime.now(UTC)
